rust/src/dcerpc/parser.rs

```diff
@@ -0,0 +1,373 @@
+
+named!(pub parse_uuid<Uuid>,
+    do_parse!(
+        time_low: take!(4) >>
+        time_mid: take!(2) >>
+        time_hi_and_version: take!(2) >>
+        clock_seq_hi_and_reserved: be_u8 >>
+        clock_seq_low: be_u8 >>
+        node: take!(6) >>
+        (
+            Uuid {
+                time_low: time_low.to_vec(),
+                time_mid: time_mid.to_vec(),
+                time_hi_and_version: time_hi_and_version.to_vec(),
+                clock_seq_hi_and_reserved: clock_seq_hi_and_reserved,
+                clock_seq_low: clock_seq_low,
+                node: node.to_vec(),
+            }
+            )
+        )
+    );
+
+fn uuid_to_vec(uuid: Uuid) -> Vec<u8> {
+    let mut uuidtmp = uuid;
+    let mut vect: Vec<u8> = Vec::new();
+    vect.append(&mut uuidtmp.time_low);
+    vect.append(&mut uuidtmp.time_mid);
+    vect.append(&mut uuidtmp.time_hi_and_version);
+    vect.push(uuidtmp.clock_seq_hi_and_reserved);
+    vect.push(uuidtmp.clock_seq_low);
+    vect.append(&mut uuidtmp.node);
+    vect
+}
+
+fn assemble_uuid(uuid: Uuid) -> Vec<u8> {
+    let mut uuidtmp = uuid;
+    let mut vect: Vec<u8> = Vec::new();
+    uuidtmp.time_low.reverse();
+    uuidtmp.time_mid.reverse();
+    uuidtmp.time_hi_and_version.reverse();
+    vect.append(&mut uuidtmp.time_low);
+    vect.append(&mut uuidtmp.time_mid);
+    vect.append(&mut uuidtmp.time_hi_and_version);
+    vect.push(uuidtmp.clock_seq_hi_and_reserved);
+    vect.push(uuidtmp.clock_seq_low);
+    vect.append(&mut uuidtmp.node);
+
+    vect
+}
+
+named!(pub dcerpc_parse_udp_header<DCERPCHdrUdp>,
+       do_parse!(
+           rpc_vers: be_u8 >>
+           pkt_type: be_u8 >>
+           flags1: be_u8 >>
+           flags2: be_u8 >>
+           drep: take!(3) >>
+           endianness: value!(if drep[0] == 0 { Endianness::Big } else { Endianness::Little }) >>
+           serial_hi: be_u8 >>
+           objectuuid: take!(16) >>
+           interfaceuuid: take!(16) >>
+           activityuuid: take!(16) >>
+           server_boot: u32!(endianness) >>
+           if_vers: u32!(endianness) >>
+           seqnum: u32!(endianness) >>
+           opnum: u16!(endianness) >>
+           ihint: u16!(endianness) >>
+           ahint: u16!(endianness) >>
+           fraglen: u16!(endianness) >>
+           fragnum: u16!(endianness) >>
+           auth_proto: be_u8 >>
+           serial_lo: be_u8 >>
+           (
+               DCERPCHdrUdp {
+                   rpc_vers: rpc_vers,
+                   pkt_type: pkt_type,
+                   flags1: flags1,
+                   flags2: flags2,
+                   drep: drep.to_vec(),
+                   serial_hi: serial_hi,
+                   objectuuid: match parse_uuid(objectuuid) {
+                       Ok((_, vect)) => assemble_uuid(vect),
+                       Err(e) => {
+                           println!("{}", e);
+                           vec![0]
+                       },
```

Comment:
What could be a better way of doing this in Rust? @jasonish ?

Reply:
This is OK, but might be better, if you expect this to fail now and again to have the objectuuid an Option..  That way you can test later if it exists or not, more idiomatically and raise an alert of whatever.

Reply:
Makes sense. Thanks.